File: src/man_ctrl/src/drive.cpp
#include <ros/ros.h>
#include <sensor_msgs/Joy.h>
#include <geometry_msgs/Twist.h>
#include <rover_msgs/WheelVelocity.h>


float x_axis_val;
float y_axis_val;
int c2,rot,rot2;



void joyCallback(const sensor_msgs::Joy::ConstPtr& joy)
{
    
    
    x_axis_val = joy->axes[0];
    y_axis_val = joy->axes[1];
  if(joy->buttons[3]==1) rot = 1;
 else if(joy->buttons[0]==1) rot =-1;
 else rot=0;
 if(joy->buttons[2]==1) rot2 = 1;
 else if(joy->buttons[2]==1) rot2 =-1;
 else rot2 =0;
 
   
   
   
    
}



int main(int argc, char** argv) {

    ros::init(argc,argv,"drive");
    ros::NodeHandle nh;
    ros::Publisher vel_pub = nh.advertise<rover_msgs::WheelVelocity>("/loco/wheel_vel",10);
    ros::Publisher cmd_pub = nh.advertise<geometry_msgs::Twist>("cmd_vel",10);
    ros::Subscriber joy_sub = nh.subscribe<sensor_msgs::Joy>("/joy",10,joyCallback);
    ros::Rate loop_rate(10);	
    


while(ros::ok())
	{

	ros::spinOnce();
    rover_msgs::WheelVelocity vel;
  geometry_msgs::Twist twist;



    c2 = 1;

    if ((y_axis_val>0.25) && (fabs(y_axis_val)>fabs(x_axis_val)))
	{
	vel.left_front_vel  = -c2 * y_axis_val*70;
        vel.right_front_vel = -c2 * y_axis_val*70; 
        vel.left_back_vel   = c2 * y_axis_val*70;
        vel.right_back_vel  = c2 * y_axis_val*70;
        twist.linear.x      =  y_axis_val;
	twist.angular.z     =   0;
     


        
	}

	
	else if ((y_axis_val< (-0.25)) && (fabs(y_axis_val)>fabs(x_axis_val)))
	{
		
	vel.left_front_vel = -c2 * y_axis_val*70;
        vel.right_front_vel= -c2 * y_axis_val*70;
        vel.left_back_vel  = c2 * y_axis_val*70;
        vel.right_back_vel = c2 * y_axis_val*70;
       twist.linear.x      =  y_axis_val;
	twist.angular.z     =   0;

	}
	
	
	else if ((x_axis_val>0.25) && (fabs(x_axis_val)>fabs(y_axis_val)))
	{
		vel.left_front_vel = -c2 * x_axis_val*70;
        vel.right_front_vel= c2 * x_axis_val*70;
        vel.left_back_vel  = -c2 * x_axis_val*70;
        vel.right_back_vel = c2 * x_axis_val*70;

        twist.linear.x      =  0;
	twist.angular.z     =   x_axis_val;
	}

	else if ((x_axis_val < (-0.25)) && (fabs(x_axis_val)>fabs(y_axis_val)))
	{
	   vel.left_front_vel = -c2 * x_axis_val*70;
       vel.right_front_vel= c2 * x_axis_val*70;
       vel.left_back_vel  = c2 * x_axis_val*70;
       vel.right_back_vel = -c2 * x_axis_val*70;

       twist.linear.x      =  0;
	twist.angular.z     =   x_axis_val;

        
	}

	else 
    {
        vel.left_front_vel = 0;
        vel.right_front_vel =  0;
        vel.left_back_vel =  0;
        vel.right_back_vel =  0;
    }
	
	vel.rot = rot;
	vel.rots =rot2;

    vel_pub.publish(vel);
    cmd_pub.publish(twist);
	loop_rate.sleep();
}
    ros::spin();
    return 0;
}

File: src/man_ctrl/scripts/drive.py
#!/usr/bin/env python
from rover_msgs.msg import WheelVelocity
from rover_msgs.msg import Diag_wheel
from sensor_msgs.msg import Joy
import math

class drive():

    def __init__(self):

        rospy.init_node("drive")

        self.pub_motor = rospy.Publisher("loco/wheel_vel",WheelVelocity,queue_size=10)
        rospy.Subscriber("diag/wheel_vel",Diag_wheel,self.Callback)
        rospy.Subscriber("/joy",Joy,self.joyCallback)

        self.theta_r = 1
        self.theta_l = 1
        self.radius_of_curve = 1
        self.chassis_width=52
        self.chassis_length = 60
        self.vtheta_r=0
        self.vtheta_l=0
        self.straight = 0
        self.zero_turn = 0
        self.steer_straight = 0
        self.steer_vel=0
        self.left_steer_vel=0
        self.right_steer_vel=0
        self.mode=0
        self.right_steer_zero = 300
        self.left_steer_zero  = 300

    def spin(self):
        rate = rospy.Rate(10)
        while not rospy.is_shutdown():
            self.main()
            rate.sleep

    def main(self):

        vel = WheelVelocity()

        if(self.mode==0):

            vel.left_front_vel  = 0
            vel.right_front_vel = 0
            vel.left_back_vel   = 0
            vel.right_back_vel  = 0

        if(self.mode==1):
            steer(self.steer_straight)
            vel.left_front_vel  = self.straight*self.vtheta_l*100
            vel.right_front_vel = self.straight*self.vtheta_r*100
            vel.left_back_vel   = self.straight*self.vtheta_l*100
            vel.right_back_vel  = self.straight*self.vtheta_r*100
            vel.right_steer     = self.right_steer_zero + (self.theta_r*10) # In the joyCallback convert the values adding the the constants also there.
            vel.left_steer      = self.left_steer_zero  + (self.theta_l*10)

        if(self.mode ==2):
            vel.left_front_vel  = -self.zero_turn*100                 # make the direction correct
            vel.right_front_vel =  self.zero_turn*100
            vel.left_back_vel   = -self.zero_turn*100
            vel.right_back_vel  =  self.zero_turn*100
        if(self.mode ==3):
            vel.right_steer_vel = self.right_steer_vel
            vel.left_steer_vel  = self.left_steer_vel

        self.pub_motor.publish(vel)


    def steer(self,value):
        if(value<0):
            self.theta_r = radians(12.5*value)
            self.theta_l = radians(20*value)
        elif(value>0):
            self.theta_r = radians(20*value)
            self.theta_l = radians(12.5*value)

        self.radius_of_curve = (self.chassis_length/(2*tan(max(abs(theta_r),abs(theta_l))))) - (self.chassis_width/2)

        self.vtheta_r = self.chassis_length/(2*sin(theta_r)*self.radius_of_curve)
        self.vtheta_l = self.chassis_length/(2*sin(theta_l)*self.radius_of_curve)



    def joyCallback(self,msg):

        self.straight  = msg.axes[1]
        self.zero_turn = msg.axes[3]
        self.steer_straight = msg.axes[2]


        if(msg.buttons[5]==1):
            self.right_steer_vel = 1
        elif(msg.buttons[7]==1):
            self.right_steer_vel = -1
        else:
            self.right_steer_vel = 0


        if(msg.buttons[4]==1):
            self.left_steer_vel = 1
        elif(msg.buttons[6]=1):
            self.left_steer_vel = -1
        else:
            self.left_steer_vel = 0


        if(msg.buttons[0]==1):
            self.mode = 1       #Normal mode
        elif(msg.buttons[3]==1):
            self.mode = 2       #Zero turn mode
        elif(msg.buttons[2]==1):
            self.mode = 3       #Calibrate mode
        elif(msg.buttons[1]==1):
            self.mode = 0       #Emergency stop mode

    def Callback(self,msg):

        self.right_steer_zero = msg.right_pot_zero
        self.left_steer_zero  = msg.left_pot_zero



if __name__ == '__main__':
    run = drive()
    run.spin()
